docker/resource_docker_service_funcs.go

```diff
@@ -0,0 +1,883 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"os"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/swarm"
+	dc "github.com/fsouza/go-dockerclient"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+var (
+	numberedStates = map[swarm.TaskState]int64{
+		swarm.TaskStateNew:       1,
+		swarm.TaskStateAllocated: 2,
+		swarm.TaskStatePending:   3,
+		swarm.TaskStateAssigned:  4,
+		swarm.TaskStateAccepted:  5,
+		swarm.TaskStatePreparing: 6,
+		swarm.TaskStateReady:     7,
+		swarm.TaskStateStarting:  8,
+		swarm.TaskStateRunning:   9,
+
+		// The following states are not actually shown in progress
+		// output, but are used internally for ordering.
+		swarm.TaskStateComplete: 10,
+		swarm.TaskStateShutdown: 11,
+		swarm.TaskStateFailed:   12,
+		swarm.TaskStateRejected: 13,
+	}
+
+	longestState int
+)
+
+type convergeConfig struct {
+	interval time.Duration
+	monitor  time.Duration
+	timeout  time.Duration
+}
+
+/////////////////
+// TF CRUD funcs
+/////////////////
+func resourceDockerServiceExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*ProviderConfig).DockerClient
+	if client == nil {
+		return false, nil
+	}
+
+	apiService, err := fetchDockerService(d.Id(), d.Get("name").(string), client)
+	if err != nil {
+		return false, err
+	}
+	if apiService == nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func resourceDockerServiceCreate(d *schema.ResourceData, meta interface{}) error {
+	var err error
+	client := meta.(*ProviderConfig).DockerClient
+
+	serviceSpec, err := createServiceSpec(d)
+	if err != nil {
+		return err
+	}
+
+	createOpts := dc.CreateServiceOptions{
+		ServiceSpec: serviceSpec,
+	}
+
+	if v, ok := d.GetOk("auth"); ok {
+		createOpts.Auth = authToServiceAuth(v.(map[string]interface{}))
+	} else {
+		createOpts.Auth = fromRegistryAuth(d.Get("image").(string), meta.(*ProviderConfig).AuthConfigs.Configs)
+	}
+
+	if v, ok := d.GetOk("update_config"); ok {
+		createOpts.UpdateConfig, _ = createUpdateOrRollbackConfig(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("rollback_config"); ok {
+		createOpts.RollbackConfig, _ = createUpdateOrRollbackConfig(v.([]interface{}))
+	}
+
+	service, err := client.CreateService(createOpts)
+	if err != nil {
+		return err
+	}
+
+	if v, ok := d.GetOk("converge_config"); ok {
+		convergeConfig := createConvergeConfig(v.([]interface{}))
+		if err := waitOnService(context.Background(), client, convergeConfig, service.ID); err != nil {
+			if err := deleteService(service.ID, d, client); err != nil {
+				return err
+			}
+			return err
+		}
+	}
```

Comment:
It seems that the service can create successfully (line 95-98 above), but then if this `converge_config` fails, we completely delete the service? I feel like we should at least check what kind of error was returned here.

Reply:
**Update:** introduced `DidNotConvergeError` which will cause the service to be deleted